src/installer/BeatPackageCompiler/AgentCustomAction.cs

```diff
@@ -0,0 +1,103 @@
+
+                System.Diagnostics.Process process = new System.Diagnostics.Process();
+                process.StartInfo.WorkingDirectory = install_folder;
+                process.StartInfo.FileName = Path.Combine(install_folder, "elastic-agent.exe");
+                process.StartInfo.Arguments = "install -f " + install_args;
+                StartProcess(session, process);
+
+                session.Log("Agent install return code:" + process.ExitCode);
+
+                if (process.ExitCode == 0)
+                {
+                    // If agent got installed properly, we can go ahead and remove all the files installed by the MSI (best effort)
+                    RemoveFolder(session, install_folder);
+                }
+
+                return process.ExitCode == 0 ? ActionResult.Success : ActionResult.Failure;
+            }
+            catch (Exception ex)
+            {
+                session.Log("Exception: " + ex.ToString());
+                return ActionResult.Failure;
+            }
+        }
+
+        private static void StartProcess(Session session, Process process)
+        {
+            // https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.process.standardoutput?view=net-8.0
+            process.StartInfo.UseShellExecute = false;
+            process.StartInfo.RedirectStandardOutput = true;
+            process.StartInfo.CreateNoWindow = true;
+            session.Log("Running command: " + process.StartInfo.FileName + " " + process.StartInfo.Arguments);
+            process.Start();
+            session.Log(process.StandardOutput.ReadToEnd());
+            process.WaitForExit();
+        }
+
+        private static void RemoveFolder(Session session, string folder)
+        {
+            try
+            {
+                new DirectoryInfo(folder).Delete(true);
+            }
+            catch (Exception ex)
+            {
+                session.Log("Failed to remove folder: " + folder + ", exception: " + ex.ToString());
+            }
+        }
+
+        [CustomAction]
+        public static ActionResult UpgradeAction(Session session)
+        {
+            session.Log("Detected an agent upgrade via MSI, which is not supported. Aborting.");
+            return ActionResult.Failure;
+        }
+
+        [CustomAction]
+        public static ActionResult UnInstallAction(Session session)
+        {
+            try
+            {
+                string install_args = string.IsNullOrEmpty(session["INSTALLARGS"]) ? "" : session["INSTALLARGS"];
+                System.Diagnostics.Process process = new System.Diagnostics.Process();
+                process.StartInfo.FileName = @"c:\\Program Files\\Elastic\\Agent\\elastic-agent.exe";
+                process.StartInfo.Arguments = "uninstall -f " + install_args;
+                StartProcess(session, process);
+
+                session.Log("Agent uninstall return code:" + process.ExitCode);
```

Comment:
We should check the exit code from uninstall, if uninstall fails, say by adding invalid installargs like `--delay-enroll` then the MSI successfully uninstalls but Agent is still on the system

Reply:
fixed